libtsuba/src/s3.cpp

```
@@ -888,4 +894,104 @@ galois::Result<void> S3DownloadRange(const std::string& bucket,
+                     object);
+  }
+  auto s3result = s3outcome.GetResult();
+  if (s3result.GetIsTruncated()) {
+    // Get more listings
+    auto continuation_token =
+        FromAwsString(s3result.GetNextContinuationToken());
+    GALOIS_LOG_VASSERT(continuation_token.length() > 0,
+                       "ListAsync IsTruncated true, but token empty");
+    s3aw.SetToken(continuation_token);
+    s3aw.Push(internal::S3ListAsyncAW);
+  }
+  auto& listing = s3aw.GetListingRef();
+  for (const auto& content : s3result.GetContents()) {
+    std::string file(FromAwsString(content.GetKey()));
+    if (file.find(object) == 0) {
+      file = std::string(file.begin() + object.size() + 1, file.end());
+    }
+    listing.emplace(file);
+  }
+
+  return galois::ResultSuccess();
+}
+
+galois::Result<std::unique_ptr<FileAsyncWork>>
+S3ListAsync(const std::string& bucket, const std::string& object) {
+  std::unique_ptr<internal::S3AsyncWork> s3aw =
+      std::make_unique<internal::S3AsyncWork>(bucket, object);
+  auto res = S3ListAsyncAW(*s3aw);
+  if (!res) {
+    GALOIS_LOG_DEBUG("S3ListAsync failure [{}]{}", bucket, object);
+    return res.error();
+  }
+  return std::unique_ptr<FileAsyncWork>(std::move(s3aw));
+}
+
+static void
+S3SendDelete(const Aws::Vector<Aws::S3::Model::ObjectIdentifier>& aws_objs,
+             const std::string& bucket) {
+  Aws::S3::Model::DeleteObjectsRequest object_request;
+  Aws::S3::Model::Delete aws_delete;
+  aws_delete.SetObjects(aws_objs);
+  object_request.SetDelete(aws_delete);
+  object_request.SetBucket(ToAwsString(bucket));
+  // fmt::print("DELETE [{}] files: {} {}\n",
+  // bucket, aws_objs.size(), (*aws_objs.begin()).GetKey());
+  auto s3_client = GetS3Client();
+  auto s3outcome = s3_client->DeleteObjects(object_request);
+  if (!s3outcome.IsSuccess()) {
+    const auto& error = s3outcome.GetError();
+    GALOIS_LOG_FATAL("\n  Failed Delete\n  {}: {}\n  [{}]",
```

Why crash instead of letting this function return an error?

---

Same as above, but you are correct, there is no reason to fail on an error from delete.  I'll change it.